Fixed #26608 -- Add support for OVER-clause (window expressions).

atombrella opened the pull request:
https://code.djangoproject.com/ticket/26608

Thanks to Jamie Cockburn for initial patch.

auvipy commented on docs/ref/contrib/postgres/window.txt:
```diff
@@ -0,0 +1,253 @@
+.. module:: django.contrib.postgres.window
+    :synopsis: Window expression API.
+
+    .. versionadded:: 2.0
```
Is this intended to be added on 1.11 or version 2.0? found the change log in 1.11 though

schinckel commented on docs/ref/contrib/postgres/window.txt:
```diff
@@ -0,0 +1,282 @@
+Window functions were introduced in ANSI SQL:2003, however support in databases
+varies.  For each backend, there is a flag that indicates support of the
+expression, and which functions that are supported in the expression, as some
+databases may allow use of all aggregation functions, and support for
```
Was there more to go here?

atombrella replied:
Yes. However, I was also adding some code for this particular part.

adamchainz commented on tests/expressions_window/models.py:
```diff
@@ -0,0 +1,18 @@
+
+class WindowTestModel(models.Model):
+    """
+    To test for Postgres window functions. The model was inspired by slides
```
rm 'postgres' 😉

adamchainz commented on docs/ref/models/window.txt:
```diff
@@ -0,0 +1,308 @@
+    the calculates the last value in a given frame clause. The type of the
+    return value is the same as ``field``.
+
+`NthValue`
```
needs double backticks

adamchainz commented on django/db/models/window.py:
```diff
@@ -0,0 +1,255 @@
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        c = self.copy()
+        return c
```
`return self.copy()`, no need for temp var

adamchainz commented on django/db/models/sql/query.py:
```diff
@@ -1487,7 +1487,7 @@ def split_exclude(self, filter_expr, prefix, can_reuse, names_with_path):
 
         It might be worth it to consider using WHERE NOT EXISTS as that has
-        saner null handling, and is easier for the backend's optimizer to
+        saner null handling, and is easier for the backend''s optimizer to
```
accident?

adamchainz commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,86 @@
+class PercentRank(Aggregate):
+    function = 'PERCENT_RANK'
+    name = 'PercentRank'
+    _output_field = FloatField()  # or DecimalField
```
is this backend specific?

atombrella replied:
No. I'm a bit in doubt what to use here.

atombrella replied:
I'm not sure what the suggested change is. However, I noticed that functions are importable from `django.db.models.functions` which I've just done (may need some additional work).

The functions can be used without an expression, and with an expression, for instance in `WITHIN GROUP`.

adamchainz commented on django/db/backends/postgresql/operations.py:
```diff
@@ -270,3 +270,28 @@ def fulltext_search_sql(self, field_name):
+        PostgreSQL disallows RANGE BETWEEN n PRECEDING AND n FOLLOWING,
+        """
+        if not isinstance(start, (int, type(None))) and isinstance(end, (int, type(None))):
+            raise TypeError('Arguments must be of type int or None')
```
You do more type checking below, you could put this error into those clauses instead of doing it twice

adamchainz commented on django/db/backends/mysql/features.py:
```diff
@@ -67,6 +69,11 @@ def is_sql_auto_is_null_enabled(self):
     @cached_property
+    def supports_over_clause(self):
+        """Only MariaDB supports this."""
+        return self.connection.mysql_version >= (10, 2)
```
I've got an idea for how we can separate MySQL/MariaDB on the DB backend coming...

adamchainz replied:
( #7778 )

charettes commented:
This is really impressive work! One thing I thought of is that filtering by an annotated window expression should be disallowed. For example:

```python
WindowTestModel.objects.annotate(sum=Window(
    expression=Sum('salary'),
    order_by=ExtractYear('hiredate'),
   frame=RowRange(start=-2, end=2))
).filter(sum__gt=100)
```

Should raise a `ValueError`.

atombrella commented:
@charettes Thank you! Also supported in Oracle, but I find PostgreSQL's presentation of the syntax a bit easier to read), your suggestion would fit neatly together with https://www.postgresql.org/docs/current/static/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS It took me a quite a while to write this initial patch, but perhaps this could be another iteration (perhaps open another ticket for these expressions).

charettes commented:
@atombrella ohh sorry about the confusion here. I was not referring to the window's expression `FILTER` clause but to attempts of referencing its alias in a `WHERE` statement.

For example, the following query is invalid AFAIK

```sql
SELECT
store_id, total,
LAG(total) OVER (PARTITION BY store_id ORDER BY timestamp DESC) as lag_total
FROM store_daily_total
WHERE lag_total > 0
```

atombrella commented:
@charettes I didn't know. I'm trying to avoid touching core-classes of the ORM, but it appears that this is needed to guard against this type of query. At least, in `resolve_expression` of `Window`, the `query.where.children` is just a blank list. If you could provide a few pointers where to look, I'd be happy to have a go at this.

About filtering, I have thought about extending it with a `FilterWindow`; MariaDB won't support this, so I guess it makes to just subclass `Window` and use a new template with `FILTER (WHERE condition)`.

charettes commented:
> @charettes I didn't know. I'm trying to avoid touching core-classes of the ORM, but it appears that this is needed to guard against this type of query. At least, in resolve_expression of Window, the query.where.children is just a blank list. If you could provide a few pointers where to look, I'd be happy to have a go at this.

I'm not exactly sure where it should live but I suppose this will require another flag on the base `Expression` flag.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,658 @@
+            ('Wilkinson', 'IT', 60000, 0.833333333333333),
+            ('Johnson', 'Management', 80000, 0.916666666666667),
+            ('Miller', 'Management', 100000, 1),
+        ], lambda row: (row.name, row.department, row.salary, row.cume_dist), ordered=False)
```
Oracle round `CumeDist` with greater then other db precision. I think that we can add `round` for test unification. 
```diff
diff --git a/tests/expressions_window/tests.py b/tests/expressions_window/tests.py
index 924928c..7982823 100644
--- a/tests/expressions_window/tests.py
+++ b/tests/expressions_window/tests.py
@@ -318,9 +318,8 @@ class TestWindowFunction(TestCase):
             expression=CumeDist(),
             order_by='salary')
         )
-
         self.assertQuerysetEqual(qs, [
-            ('Moore', 'IT', 34000, 0.0833333333333333),
+            ('Moore', 'IT', 34000, 0.083333333333333),
             ('Williams', 'Accounting', 37000, 0.166666666666667),
             ('Smith', 'Marketing', 38000, 0.25),
             ('Johnson', 'Marketing', 40000, 0.333333333333333),
@@ -332,7 +331,7 @@ class TestWindowFunction(TestCase):
             ('Wilkinson', 'IT', 60000, 0.833333333333333),
             ('Johnson', 'Management', 80000, 0.916666666666667),
             ('Miller', 'Management', 100000, 1),
-        ], lambda row: (row.name, row.department, row.salary, row.cume_dist), ordered=False)
+        ], lambda row: (row.name, row.department, row.salary, round(row.cume_dist, 15)), ordered=False)
 
     @skipUnlessSupportsFunction('NthValue')
     def test_nthvalue(self):
```

adamchainz replied:
Maybe with a comment that the `round` is for Oracle 😉

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,658 @@
+        qs = WindowTestModel.objects.annotate(dense_rank=Window(
+            expression=DenseRank(),
+            partition_by=F('department'),
+            order_by='-salary')
```
We have two employees in the same department (_"Accounting"_) and with the same salary (_"45000"_) i.e. _"Jenson_", _"Johnson_". Result is non-deterministic, it depends on the db backend default ordering, that seems to be different for each database. IMO the best solution is to add option to pass multiple field names (like in [`models/query.py`](https://github.com/django/django/blob/master/django/db/models/query.py#L926)) in `order_by` attribute, e.g.
```python
    def test_dense_rank(self):
        qs = WindowTestModel.objects.annotate(dense_rank=Window(
            expression=DenseRank(),
            partition_by=F('department'),
            order_by=['-salary', 'name'])
        ).order_by('department')
```
Ordering is also an issue in others tests.

holvianssi commented on django/db/backends/base/features.py:
```diff
@@ -227,6 +227,17 @@ class BaseDatabaseFeatures:
+
+    # Not every aggregate function is necessarily supported in every database
+    # backend
+    supported_window_aggregates = []
```
If I'm not mistaken these aren't checked anywhere yet.

I'm wondering will this style make it harder than necessary to write new window functions - if you write a new window function for your own use, will you need to then also patch the backend?

atombrella replied:
In the test suite because of MariaDB that only implements partial support for the functions.

I suppose annotations instead would be a cleaner solution, and deleting these flags.

atombrella commented:
buildbot, test on oracle

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

atombrella commented:
@jarshwah There are things that are likely not to change more (for instances the actual functions); I'm adjusting some tests because of Oracle. Aggregation over the result of a window function causes problems on Oracle because the ORM adds a group by to the inner query; I'm not sure if this is correct behavior (I had a look in `get_aggregation` and `aggregate` in `Query` and `QuerySet`, respectively, but it's not trivial). The patch is fairly big, and it's been easier to get some suggestions along the way.

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -406,23 +407,25 @@ def test_ntile(self):
+            ('Johnson', 'Management', 80000, 0.90909091),
             ('Miller', 'Management', 100000, 1.0),
-        ], transform=lambda row: (row.name, row.department, row.salary, row.percent_rank),)
+        ], transform=lambda row: (row.name, row.department, row.salary, round(row.percent_rank, 8)),)
```
Please round to the greatest common decimal precision i.e. `15`:
```diff
diff --git a/tests/expressions_window/tests.py b/tests/expressions_window/tests.py
index 70cc48c..e8fb660 100644
--- a/tests/expressions_window/tests.py
+++ b/tests/expressions_window/tests.py
@@ -411,7 +411,7 @@ class TestWindowFunction(TestCase):
 
         self.assertQuerysetEqual(qs, [
             ('Moore', 'IT', 34000, 0.0),
-            ('Williams', 'Accounting', 37000, 0.0909090909090909),
+            ('Williams', 'Accounting', 37000, 0.090909090909091),
             ('Smith', 'Marketing', 38000, 0.181818181818182),
             ('Johnson', 'Marketing', 40000, 0.272727272727273),
             ('Jenson', 'Accounting', 45000, 0.363636363636364),
@@ -422,7 +422,7 @@ class TestWindowFunction(TestCase):
             ('Wilkinson', 'IT', 60000, 0.818181818181818),
             ('Johnson', 'Management', 80000, 0.909090909090909),
             ('Miller', 'Management', 100000, 1.0),
-        ], transform=lambda row: (row.name, row.department, row.salary, row.percent_rank),)
+        ], transform=lambda row: (row.name, row.department, row.salary, round(row.percent_rank, 15)))
```

atombrella replied:
Thanks. Adjusted and pushed. When you started the tests on Oracle, it was possible to see the test failures. Could you please do that again? There are a couple of tests (aggregation and the `RANGE BETWEEN n PRECEDING AND n FOLLOWING` that still needs to be filled out). Again, I really appreciate the input/help :)

felixxm commented:
buildbot, test on oracle.

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,710 @@
+            ('Miller', 'Management', 100000, 1.0),
+        ], transform=lambda row: (row.name, row.department, row.salary, round(row.percent_rank, 15)),)
+
+    def test_nth_returns_null(self):
```
In the `test_nthvalue` Oracle returns empty string `''` instead of `None` for char fields.

atombrella replied:
OK. I changed the test to use `salary` instead. Perhaps an Oracle-specific test should be introduced to show this scenario with null/the empty string? Or simply a remark about this in the documentation, to highlight this fact. I don't think it'd be correct to use `convert_value`.

felixxm replied:
It's weird, because Oracle interprets empty strings as nulls.

atombrella replied:
This test uses an integer-field, but the name may be misleading because of this empty string/null. I've added a remark about the empty strings/null for `nth_value` in the documentation; I don't want to trigger a build just for this commit. For `lead` and `lag`, the default null is passed explicitly in the constructor.

atombrella commented:
buildbot, test on oracle.

jarshwah commented:
How are you going with this patch @atombrella ? Is there anything you're still stuck on?

atombrella commented:
I am not really sure what specifically in the code-section that adds additional work, probably some clean-up, for instance syntax for the partition and order by part. The last two commits were mostly clean-up. Is the structure of the documentation fine (its own section, functions documented along with the other functions)? It needs to be looked over by a native speaker.  @jarswah You brought up aggregation as a consideration for window functions on the mailing list; I think distinct needs to be used in most scenarios, but this is already disallowed (although postgresql supports it if you comment out the raise exception in the ORM). On the mailing list, ExpressionList was proposed for ordering. I think it's a good idea for non-ambiguous ordering clauses, as I don't see how else to add descending and ascending specifiers to expressions. There are differences among supported features in all of the backends, I think. MariaDB (although Django does not officially support it, I think this is to be changed) has limitations on OrderBy (nulls first/last is not supported) and some functions (the test suite skips those), PostgreSQL doesn't support RESPECT/IGNORE NULLS or FROM FIRST/LAST. Would these restrictions be good to highlight in the documentation (that's the current approach), or is it needed to enforce it more strictly?

atombrella commented on docs/ref/models/database-functions.txt:
```diff
@@ -786,3 +786,157 @@ Usage example::
+
+    (rank - 1) / (total rows - 1)
+
+The following example explains the calculation for the percentile rank of a
```
The table was copied from Redshift's documentation.

atombrella commented on django/db/models/expressions.py:
```diff
@@ -126,6 +126,14 @@ class BaseExpression:
+    # flag to indicate if it's possible to use the expression in an update-query
+    # contains_aggregate will add a group by if contains_aggregate which may be
+    # incorrect for some expressions
+    can_be_reffed_in_update = True
```
I'm not sure if this is necessary, or if the `for_save` argument implies this.

jarshwah replied:
`for_save` is True for inserts and updates, so you can remove this and put the fielderror into the resolve_expressions that make sense.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
@jarshwah Any input to this, e.g., are there more restrictions than the filtering on the result of a window and that they can't be used in `UPDATE`? I've tried to tidy up in a few places. Thank you.

jarshwah commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,755 @@
+        window functions.
+        """
+        WindowTestModel.objects.create(name='Jones', salary=45000, department='Accounting',
+                                       hiredate=datetime.datetime(2005, 11, 1))
```
We usually avoid hanging indents, and prefer this:

```
WindowTestModel.objects.create(
    name='Jones', salary=45000, department='Accounting', hiredate=datetime.datetime(2005, 11, 1)
)
```

jarshwah replied:
Or split the args over two lines if it passes 119 chars.

jarshwah commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,755 @@
+
+
+@skipUnlessDBFeature('supports_over_clause')
+class TestWindowFunction(TestCase):
```
Do you think there might be too much test data in use here? Without a solid understanding of window functions, it becomes quite difficult to spot what the output should be, and to pick out any potential issues. You might have carefully chosen all the test data based on all the various window functions that need to be tested, so don't just change it because someone unfamiliar can't understand them.

It'd be easier to review (and to verify in the future) if the test data was smaller, and there was a one line explanation on what the results should be. For example, the `Lead` expression might say:

> Display the salary of the person in the next row, sorted by hiredate

That would be almost duplication of the documentation, but would quickly allow verification of the results without referring back to big paragraphs of text. If it's possible to keep a short explanation in one sentence (like above) I feel that would greatly help review.

atombrella replied:
The test data can most likely be simplified. I was aiming at something that can be partitioned and ordered in different ways. Probably quite useful to have some notes about the expected output :+1:

atombrella replied:
There are a few window functions which take an `offset` argument, and thus at least one department with more than 2 employees is needed. I'm also a bit uncertain if all databases compute the same `ntile` if the number of rows in a partition does not divide the number of buckets evenly.

Also, tests for frames (which should be cleaned up) require a reasonable number of rows to make sense.

atombrella commented:
@charettes You may have some more input, e.g., `contains_aggregate = True` takes care of a lot of the same restrictions on the expression, however disallowing `DISTINCT ON` with PostgreSQL. Currently, it's false.

charettes commented on django/db/models/sql/query.py:
```diff
@@ -1156,6 +1156,10 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
+
+        if hasattr(reffed_expression, 'filterable') and not reffed_expression.filterable:
+            raise NotSupportedError(reffed_expression.__class__.__name__ + ' is disallowed in the filter-clause')
+
```
There's some overlap with #8119 here. @schinckel what do you think about making `Expression.filterable` a property that returns whether or not `isinstance(self.output_field, BooleanField)` and use it both cases?

charettes replied:
Actually the `filterable` attribute would have a different meaning here, scratch that.

charettes commented:
@atombrella about the `contains_aggregate` issue. I think that introducing a new `contains_window = True` (or `contains_frame`, `contains_over_clause`?) flag (with a default nested lookup on `BaseExpression`) that the ORM can use to branch on is the way to go here. Window functions are new concept for the ORM so it's normal we have to deal with these expressions with special care just like we do with aggregates.

jarshwah commented on django/db/models/window.py:
```diff
@@ -0,0 +1,203 @@
+            window_sql.extend(sql_expr)
+            window_params.extend(sql_params)
+
+        if self.order_by is not None:
```
Reminder to self: check if sqli is possible here. It doesn't *look* like it is, but I want to double check.

jarshwah commented on django/db/models/window.py:
```diff
@@ -0,0 +1,203 @@
+    def __repr__(self):
+        return '<%s: %s>' % (self.__class__.__name__, self)
+
+    def copy(self):
```
You can probably drop this copy if `get/set_source_expressions` is implemented instead

jarshwah commented on django/db/models/window.py:
```diff
@@ -0,0 +1,203 @@
+
+    def get_group_by_cols(self):
+        # it may be considered that expressions in the ordering clause
+        # should be referenced here
```
Drop the comment I think - it should always be empty.

atombrella commented on django/db/backends/base/features.py:
```diff
@@ -229,6 +229,9 @@ class BaseDatabaseFeatures:
     # Does the backend support indexing a TextField?
     supports_index_on_text_field = True
+    # Window expressions (aggregate OVER (expression)), currently only
+    # supported it in PostgreSQL and Oracle
```
Well, MariaDB 10.2. I think an official backend is on the way from @adamchainz

adamchainz replied:
Well, mariadb support in the mysql backend. Will get on to that soonish.

atombrella commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,120 @@
+class FirstLastValueWrapper(Aggregate):
+    def __init__(self, expression, **extra):
+        if expression is None:
+            raise ValueError('%s requires a non-null source_expression' % self.__class__.__name__)
```
I think this is the only place where `arity` makes sense. For the other functions, some of the arguments are optional which just raises an exception.

atombrella commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,120 @@
+]
+
+
+class CumeDist(Aggregate):
```
A few of them can be used in `WITHIN GROUP`, and thus fall under the category "ordered set aggregates".

https://www.postgresql.org/docs/current/static/functions-aggregate.html#FUNCTIONS-HYPOTHETICAL-TABLE

atombrella replied:
I decided to derive them all from `Func` because they're not aggregate functions in the same as `Count` and `Sum` which don't need to be part of an expression such as `WITHIN GROUP` or `OVER`. The `GROUP BY` required by the aggregate functions is also not needed in many cases, so I think your initial verdict was right.

atombrella commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,120 @@
+    name = 'PercentRank'
+
+    def __init__(self, *expressions, **extra):
+        super().__init__(*expressions, output_field=FloatField())
```
Warnings about `DEFAULT_INDEX_TABLESPACE` was the reason for `__init__`. I found it a bit peculiar why it worked with __init__ and not with the attribute.

It seems the standard is to have functions importable from `django.db.models.functions`.

atombrella replied:
I'd like some guidance how to avoid the `__init__`-method. You can reproduce the described warning simply by adding `_output_field = FloatField()` (or `IntegerField()`), and then dropping the `__init__`-methods.

jarshwah replied:
It seems I was wrong - there are so many examples in the code that override init just to force the output field it's not worth worrying about here. Maybe in the future we can make it easier, but not necessary for this patch.

charettes replied:
Defining `__init__()` to specify a default `output_field` is required until we re-arrange `runtests.py` to avoid importing settings dependant modules before calling `django.setup()`.

atombrella commented on django/db/models/expressions.py:
```diff
@@ -749,6 +761,61 @@ def get_group_by_cols(self):
         return [self]
 
 
+class ExpressionList(Expression):
```
`__repr__` and `__str__` are also implemented. I'm not sure if that's ideal, it made it easier in `Window` to write the `__str__` and `__repr__`-methods.

atombrella commented:
@jarshwah Thanks, the comments were quite useful. For convenience, I summarized the changes in the last commit message (there are still a few things that you commented that I'd like to dig into, such as the tagging of the compatible functions). About syntax for `order_by` and `partition_by`, I think a list that involves `ExpressionList` only internally in the `Window`-class could be an idea.

Empty windows are allowed, the three optional arguments should just be left empty. I can submit a ticket about `filter` and `Window` expressions later.

atombrella commented:
buildbot, test on oracle.

jarshwah commented:
So before this should go through another round of review the problem of window expressions causing a GROUP BY to be added to the query needs to be addressed. I'm going to move the ticket back into patch needs improvement.

Also, Aggregates should not inherit from WindowFunc. Just add the window_compatible property to the base aggregate type. I noticed that some of the docs around the window_compatible property need to be cleaned up.

Once the GROUP BY issue is sorted I can give another review. Can you ping me here when that's done, and I'll make it a priority.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
http://mysqlserverteam.com/mysql-8-0-2-introducing-window-functions/ <- I haven't considered this in the current patch (but it affects just the test annotations and the feature-flag). I think `window_compatible` should remain `True` for all aggregate functions, although support is limited to just a few of them on MySQL and MariaDB.

jarshwah commented:
I'll be finishing off the review for this at the pycon au sprints this coming Monday

MarkusH commented on django/db/models/expressions.py:
```diff
@@ -748,6 +765,32 @@ def get_group_by_cols(self):
+        return '<{}: {}>'.format(self.__class__.__name__, self.__str__())
+
+    def __str__(self):
+        return '{args}'.format(
```
How about this
```python
def __str__(self):
    return self.arg_joiner.join(str(arg) for arg in self.source_expressions)
```

I don't see a need for a `.format()` call.

MarkusH commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,119 @@
+    name = 'CumeDist'
+
+    def __init__(self, *expressions, **extra):
+        super().__init__(*expressions, output_field=FloatField())
```
This is a recurring thing occurring here. What do you think about giving `Expression` a `get_output_field()` method and a `output_field` attribute that it can look up, @jarshwah ?

jarshwah replied:
See https://github.com/django/django/pull/7611/files#r131567089

We should make this easier - but in a separate patch. There's already a def output_field and def _output_field.. we can fix it all up at once.

charettes replied:
`ouput_field` should work just fine, the issue is really that `runtests.py` should follow our documented guidelines and make sure to call `django.setup()` before attempting any other `django` imports. Here's the stacktrace @atombrella mentioned (triggered by setting `django.db.models.functions.base.Now.output_field = fields.DateTimeField()` and trying to run tests).

```
Traceback (most recent call last):
  File "./runtests.py", line 17, in <module>
    from django.test import TestCase, TransactionTestCase
  File "/Users/charettes/Workspace/django/django/test/__init__.py", line 3, in <module>
    from django.test.client import Client, RequestFactory
  File "/Users/charettes/Workspace/django/django/test/client.py", line 12, in <module>
    from django.core.handlers.base import BaseHandler
  File "/Users/charettes/Workspace/django/django/core/handlers/base.py", line 7, in <module>
    from django.urls import get_resolver, set_urlconf
  File "/Users/charettes/Workspace/django/django/urls/__init__.py", line 1, in <module>
    from .base import (
  File "/Users/charettes/Workspace/django/django/urls/base.py", line 9, in <module>
    from .exceptions import NoReverseMatch, Resolver404
  File "/Users/charettes/Workspace/django/django/urls/exceptions.py", line 1, in <module>
    from django.http import Http404
  File "/Users/charettes/Workspace/django/django/http/__init__.py", line 5, in <module>
    from django.http.response import (
  File "/Users/charettes/Workspace/django/django/http/response.py", line 14, in <module>
    from django.core.serializers.json import DjangoJSONEncoder
  File "/Users/charettes/Workspace/django/django/core/serializers/__init__.py", line 23, in <module>
    from django.core.serializers.base import SerializerDoesNotExist
  File "/Users/charettes/Workspace/django/django/core/serializers/base.py", line 6, in <module>
    from django.db import models
  File "/Users/charettes/Workspace/django/django/db/models/__init__.py", line 19, in <module>
    from django.db.models.manager import Manager
  File "/Users/charettes/Workspace/django/django/db/models/manager.py", line 6, in <module>
    from django.db.models.query import QuerySet
  File "/Users/charettes/Workspace/django/django/db/models/query.py", line 23, in <module>
    from django.db.models.functions import Trunc
  File "/Users/charettes/Workspace/django/django/db/models/functions/__init__.py", line 1, in <module>
    from .base import (
  File "/Users/charettes/Workspace/django/django/db/models/functions/base.py", line 161, in <module>
    class Now(Func):
  File "/Users/charettes/Workspace/django/django/db/models/functions/base.py", line 163, in Now
    output_field = fields.DateTimeField()
  File "/Users/charettes/Workspace/django/django/db/models/fields/__init__.py", line 1161, in __init__
    super().__init__(verbose_name, name, **kwargs)
  File "/Users/charettes/Workspace/django/django/db/models/fields/__init__.py", line 161, in __init__
    self.db_tablespace = db_tablespace or settings.DEFAULT_INDEX_TABLESPACE
  File "/Users/charettes/Workspace/django/django/conf/__init__.py", line 56, in __getattr__
    self._setup(name)
  File "/Users/charettes/Workspace/django/django/conf/__init__.py", line 41, in _setup
    % (desc, ENVIRONMENT_VARIABLE))
django.core.exceptions.ImproperlyConfigured: Requested setting DEFAULT_INDEX_TABLESPACE, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
```

MarkusH commented on docs/ref/models/window.txt:
```diff
@@ -0,0 +1,134 @@
+        >>> from django.db.models import Avg, ExpressionList, F, Window
+        >>>
+        >>> Movie.objects.prefetch_related('director').annotate(
+        >>>     avg_rating=Window(
```
Can you put the closing `)` on the same indent level as `avg_rating` please.

MarkusH commented on docs/ref/models/window.txt:
```diff
@@ -0,0 +1,134 @@
+
+      >>> from django.db.models import F, RowRange, Window
+      >>>
+      >>> Book.objects.prefetch_related('author').annotate(avg_rating=Window(
```
Can you add a line break before `avg_rating=` please. I.e. follow the same indentation design as above.

MarkusH commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,880 @@
+        PostgreSQL, and from some third-party material that explained
+        window functions.
+        """
+        WindowTestModel.objects.create(
```
Since you don't actually need the instances of all those models, can you use `WindowTestModel.objects.bulk_create()` please.

jarshwah commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,119 @@
+]
+
+
+class WindowFunction(Func):
```
I think having a separate base class for setting a single flag is probably overkill. If there are future changes that require more customisation then we can introduce a base class to keep things tidy. In the mean time let's not.

jarshwah commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,119 @@
+class Ntile(WindowFunction):
+    function = 'NTILE'
+    name = 'Ntile'
+    arity = 1
```
I might have suggested it, but I don't think arity is useful in this type.

jarshwah commented on django/db/models/window.py:
```diff
@@ -0,0 +1,189 @@
+
+        if self.order_by is not None:
+            if not isinstance(self.order_by, (OrderBy, Expression, list)):
+                raise ValueError('order_by must either be of type OrderBy or an Expression')
```
Error message doesn't match the code, and we're ignoring the possibility of tuples. Not to mention that OrderBy *is* an Expression.

What do we want to happen here? If it's an expression, do we need to order it (by calling asc/desc)? Or do we just want to push any expression into an ExpressionList?

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

adamchainz commented on django/db/models/window.py:
```diff
@@ -0,0 +1,190 @@
+                self.partition_by = ExpressionList(self.partition_by)
+
+        if self.order_by is not None:
+            if not isinstance(self.order_by, (BaseExpression, OrderBy, list, tuple)):
```
You can remove `OrderBy` from here because it's a subclass of `BaseExpression`. Also could use `collections.Sequence` instead of `list` and `tuple` to allow anything that the * operator will work with.

The message could also read 'must be either an Expression or a sequence of expresssions'

atombrella replied:
`Sequence` also covers `str`, so without typing annotations, I'd like to postpone this suggestion till later. `str` is specified as an invalid argument type (please see `test_invalid_order_by`).

adamchainz replied:
Ok fair enough!

sir-sigurd commented on django/db/models/window.py:
```diff
@@ -0,0 +1,190 @@
+        self.source_expression = self._parse_expressions(expression)[0]
+
+    @cached_property
+    def contains_aggregate(self):
```
I think `contains_aggregate` and `contains_over_clause` could be defined as class attributes for simplicity.

sir-sigurd commented on django/db/backends/base/operations.py:
```diff
@@ -598,3 +607,39 @@ def subtract_temporals(self, internal_type, lhs, rhs):
+        if not self.connection.features.supports_over_clause:
+            raise NotImplementedError('This backend does not support window expressions.')
+
+        if isinstance(start, int):
```
Might be better to define this as separate method like this:
```
def window_frame_start(self, start):
    if isinstance(start, int):
        if start < 0:
            return '%d %s' % (abs(start), self.PRECEDING)
        elif start == 0:
            return self.CURRENT_ROW
    elif start is None:
        return self.UNBOUNDED_PRECEDING
    raise ValueError('Illegal argument for start, must be either a negative integer, zero or None')
```

sir-sigurd commented on django/db/models/window.py:
```diff
@@ -0,0 +1,190 @@
+        Oracle and PostgreSQL.
+        """
+        connection.ops.check_expression_support(self)
+        if self.frametype.lower() == 'rows':
```
Might be worth to define method `window_frame_start_end(connection, start, end)` for `RowRange` and `ValueRange` and call it unconditionally here.

atombrella replied:
Yes. I'll push an update with this later. Thanks :)

sir-sigurd commented on django/db/models/window.py:
```diff
@@ -0,0 +1,190 @@
+        return clone
+
+    def set_source_expressions(self, exprs):
+        self.start, self.end = exprs[0], exprs[1]
```
Simplify to `self.start, self.end = exprs`?

sir-sigurd commented on django/db/models/window.py:
```diff
@@ -0,0 +1,190 @@
+        self.start, self.end = exprs[0], exprs[1]
+
+    def get_source_expressions(self):
+        return (self.start, self.end),
```
Return list for consistency?

atombrella commented:
buildbot, test on oracle.

atombrella commented:
@timgraham Would you have time to review docs soon? Thank you. I think the changes in `database-functions.txt` are mostly stand-alone.

felixxm commented on django/db/models/expressions.py:
```diff
@@ -140,6 +140,12 @@ class BaseExpression:
     is_summary = False
     _output_field_resolved_to_none = False
 
+    # If it can be used in a WHERE-clause
```
Maybe `# If the expression can be used in a WHERE-clause.`.

felixxm commented on django/db/models/expressions.py:
```diff
@@ -140,6 +140,12 @@ class BaseExpression:
+    # If it can be used in a WHERE-clause
+    filterable = True
+
+    # If the expression can be used as source expression in Window
```
Dot is missing.
`as source expression` -> `as a source expression`.

felixxm commented on django/db/models/expressions.py:
```diff
@@ -479,6 +493,9 @@ def as_sql(self, compiler, connection):
 @deconstructible
 class F(Combinable):
     """An object capable of resolving references to existing query objects."""
+    # If it can be used in a WHERE-clause
```
Maybe `# If the expression can be used in a WHERE-clause.`.

felixxm commented on django/db/models/expressions.py:
```diff
@@ -765,6 +782,30 @@ def get_group_by_cols(self):
+    ordering clause.
+    """
+    template = '%(expressions)s'
+    arg_joiner = ', '
```
This line can be removed since `Func` already has `arg_joiner`.

felixxm commented on django/db/models/expressions.py:
```diff
@@ -765,6 +782,30 @@ def get_group_by_cols(self):
+        extra = ', '.join(str(key) + '=' + str(val) for key, val in self.extra.items())
+        if extra:
+            return '<{}: {}, {}>'.format(self.__class__.__name__, self.__str__(), extra)
+        return '<{}: {}>'.format(self.__class__.__name__, self.__str__())
```
`ExpressionList.__repr__()` can be removed because `Func.__repr__()` works in the same way.

felixxm commented on django/db/models/sql/compiler.py:
```diff
@@ -1110,6 +1110,8 @@ def prepare_value(self, field, value):
             if value.contains_aggregate:
                 raise FieldError("Aggregate functions are not allowed in this query")
+            if value.contains_over_clause:
+                raise FieldError('Window expressions are not allowed in this query')
```
Dot is missing. Please add a dot to all exception messages.

atombrella replied:
I just copied the pattern from the line above. I'll adjust the newly introduced messages in this PR.

felixxm commented on django/db/models/window.py:
```diff
@@ -0,0 +1,187 @@
+    def __init__(self, start=None, end=None):
+        self.start = start
+        self.end = end
+        super().__init__(output_field=None)
```
I think you don't need to pass `output_field=None` to the `Expression.__init__()`.

atombrella replied:
I don't think this `__init__`-call is even needed.

felixxm commented on django/db/models/window.py:
```diff
@@ -0,0 +1,187 @@
+
+    def __str__(self):
+        if self.start is not None and self.start < 0:
+            start = '%d PRECEDING' % abs(self.start)
```
I think you should use constants from `DatabaseOperations` instead of `'PRECEDING'`, `'CURRENT ROW'`, `'UNBOUNDED'` etc.

felixxm commented on docs/ref/models/database-functions.txt:
```diff
@@ -819,3 +819,145 @@ Usage example::
+    evaluated with respect to the current row.
+
+    No validation is done on this expression, i.e., the types of
+    ``expression`` and ``default`` are not compared.
```
Please wrap docs at 79 characters, i.e.:
```
    Calculate the leading value in a given frame. The fallback ``default`` must
    be of the same type as ``element``. Both offset and default are evaluated
    with respect to the current row.

    No validation is done on this expression, i.e., the types of ``expression``
    and ``default`` are not compared.
```

atombrella replied:
I formatted the other doc-file, but clearly forgot this. Thanks.

felixxm commented on tests/db_functions/test_window_functions.py:
```diff
@@ -0,0 +1,35 @@
+from django.test import TestCase, skipUnlessDBFeature
+
+
+@skipUnlessDBFeature('supports_over_clause')  # no need to run on mysql or sqlite
```
I think this comment (`# no need to run on mysql or sqlite`) is unnecessary.

felixxm commented on tests/expressions/tests.py:
```diff
@@ -1351,6 +1355,12 @@ def test_expressions(self):
+            '<ExpressionList: F(col), F(anothercol)>')
+        self.assertEqual(
+            repr(ExpressionList(OrderBy(F('col'), descending=False))),
+            '<ExpressionList: OrderBy(F(col), descending=False)>')
```
Please use hanging indentation as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style)

felixxm commented on tests/expressions_window/models.py:
```diff
@@ -0,0 +1,18 @@
+    hiredate = models.DateField(blank=False, null=False)
+
+    def __str__(self):
+        return "{}, {}, {}, {}".format(self.name, self.department, self.salary, self.hiredate)
```
Use single quotes.

dekoza commented on django/db/backends/base/operations.py:
```diff
@@ -598,3 +607,43 @@ def subtract_temporals(self, internal_type, lhs, rhs):
+        start = self.window_frame_start(start)
+        end = self.window_frame_end(end)
+
+        return start, end
```
It is helpful when using (i)pdb.

adamchainz replied:
Fewer temporary variables make it go faster, I agree with felix

dekoza commented on django/db/backends/postgresql/operations.py:
```diff
@@ -247,3 +248,10 @@ def subtract_temporals(self, internal_type, lhs, rhs):
+
+    def window_frame_range_start_end(self, start=None, end=None):
+        start_, end_ = super().window_frame_range_start_end(start, end)
+        if (start and start < 0) and (end and end > 0):
```
> In the face of ambiguity refuse the temptation to guess

atombrella commented on django/db/backends/base/operations.py:
```diff
@@ -598,3 +607,43 @@ def subtract_temporals(self, internal_type, lhs, rhs):
+            elif start == 0:
+                return self.CURRENT_ROW
+            else:
+                raise ValueError('start must be a negative integer, zero or None')
```
Good. This also makes two tests unneeded.

atombrella commented on django/db/backends/base/operations.py:
```diff
@@ -598,3 +607,43 @@ def subtract_temporals(self, internal_type, lhs, rhs):
+        Return corresponding text for use in an OVER-clause window-frame.
+        """
+        if not self.connection.features.supports_over_clause:
+            raise NotImplementedError('This backend does not support window expressions.')
```
The usual pattern in that file uses `This backend ...`, and `NotImplementedError`.

atombrella commented:
buildbot, test on oracle.

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,866 @@
+        qs = WindowTestModel.objects.annotate(max_salary_year=Window(
+            expression=Max('salary'),
+            order_by=ExtractYear('hiredate').asc(),
+            partition_by=ExtractYear('hiredate'))
```
The last parenthesis should be moved to the next line (as in the `test_dense_rank`).

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,866 @@
+        """
+        qs = WindowTestModel.objects.annotate(cume_dist=Window(
+            expression=CumeDist(),
+            order_by=F('salary').asc())
```
The last parenthesis should be moved to the next line (as in the `test_dense_rank`).

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,866 @@
+            nth_value=Window(expression=NthValue(
+                expression='salary', nth=2),
+                order_by=[F('hiredate').asc(), F('name').desc()],
+                partition_by=F('department'))
```
The last parenthesis should be moved to the next line (as in the `test_dense_rank`). Please check also other tests.

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,866 @@
+            ('Smith', 38000, 'Marketing', 60000),
+            ('Brown', 53000, 'Sales', 60000),
+            ('Smith', 55000, 'Sales', 60000),
+        ], transform=lambda row: (row.name, row.salary, row.department, row.lead_default), ordered=False)
```
I don't think we need to check all rows, probably sth like this:
```python
self.assertEqual(list(qs.values_list('lead_default', flat=True).distinct()), [60000])
```
will be sufficient. We have a similar situation in the `test_nth_returns_null`.

atombrella commented on docs/ref/models/window.txt:
```diff
@@ -0,0 +1,175 @@
+================
```
I think I paid more attention to just writing documentation, and not where to put it, and put it in this file. I'm very neutral as to where the text is located. `expressions.txt` seems like a good place.

atombrella commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,885 @@
+        """
+        WindowTestModel.objects.bulk_create([
+            WindowTestModel(
+                name='Jones', salary=45000, department='Accounting', hiredate=datetime.datetime(2005, 11, 1),
```
I like that idea. Thanks. It reduced the file by about 20 lines.

atombrella commented on tests/expressions_window/models.py:
```diff
@@ -0,0 +1,18 @@
+
+class WindowTestModel(models.Model):
+    """
+    To test for window functions. The model was inspired by slides explaining
```
Comment is deleted.

atombrella commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,866 @@
+    def test_invalid_filter(self):
+        msg = 'Window is disallowed in the filter-clause'
+        with self.assertRaisesMessage(NotSupportedError, msg):
+            WindowTestModel.objects.annotate(dense_rank=Window(
```
This example works, but adapting it to `test_nth_returns_null`, I get a list of `[None]` x 12. I checked with the corresponding SQL in PostgreSQL, and it returns only one null-value. Is this an edge-case worth reporting a ticket for?

felixxm replied:
I've opened a separate [ticket](https://code.djangoproject.com/ticket/28560) for that.

atombrella commented on django/db/backends/mysql/features.py:
```diff
@@ -83,6 +83,11 @@ def is_sql_auto_is_null_enabled(self):
 
     @cached_property
+    def supports_over_clause(self):
+        """Only MariaDB 10.2+ and MySQL 8.0.2+ support this."""
```
I haven't run the test suite on MySQL 8 (yet). It's still in unstable mode. I removed test annotations for MariaDB, but will try to run the test suite on 8.0.2 before pushing an update.

atombrella commented:
@timgraham Thanks for the comments. I adjusted the annotations in the tests, and experimented a bit with a pre-release of MySQL 8; they all ran, except for one. I adjusted the `test_range_n_preceding_and_following` due to this. Would it be an idea to highlight something about experimental support/tests for MySQL in the documentation? The tests run on MySQL 8.0.2, but MariaDB still has hiccups with lead/lag, and `range` (MariaDB should have `lead` and `lag`, but the default argument resulted in an error about an invalid query).

As the implementation uses the `Expression` API, I find it useful to highlight a trick with a dictionary for the parameters for multiple windows in the same query. I moved the document `window.txt` into `expressions.txt`. Most sections start with `.. class`, and then some text, and I'm not sure if the section requires some structural change.

The arguments for the frame were taken from sqlalchemy; the default in SQL is that `RANGE/ROWS UNBOUNDED PRECEDING` means `RANGE/ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`, although without an elaborate end, the code produces `RANGE/ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING`, which I've tried to highlight.

atombrella commented:
buildbot, test on oracle.

adamchainz commented:
>  Would it be an idea to highlight something about experimental support/tests for MySQL in the documentation?

Thanks for testing MySQL :) I think a sentence about MySQL 8 support would be okay, even though it's currently unreleased.

> MariaDB still has hiccups with lead/lag, and range (MariaDB should have lead and lag, but the default argument resulted in an error about an invalid query).

Don't worry about MariaDB for the time being, I'll try make it supported properly by Django and fix up things like this when I get round to it.

timgraham commented on django/db/backends/mysql/features.py:
```diff
@@ -83,6 +83,11 @@ def is_sql_auto_is_null_enabled(self):
 
     @cached_property
+    def supports_over_clause(self):
+        """Only MySQL 8.0.2+ support this."""
```
I don't think the docstring is needed -- the comparison is straightforward.

timgraham commented on django/db/backends/base/operations.py:
```diff
@@ -598,3 +605,35 @@ def subtract_temporals(self, internal_type, lhs, rhs):
+                return self.CURRENT_ROW
+        elif start is None:
+            return self.UNBOUNDED_PRECEDING
+        raise ValueError('Illegal argument for start, must be either a negative integer, zero or None.')
```
"start argument must be a negative integer, zero, or None, but got %s."

timgraham commented on django/db/backends/base/operations.py:
```diff
@@ -598,3 +605,35 @@ def subtract_temporals(self, internal_type, lhs, rhs):
+
+    def window_frame_rows_start_end(self, start=None, end=None):
+        """
+        Return corresponding text for use in an OVER-clause window-frame.
```
"corresponding" to what?

timgraham commented on django/db/models/expressions.py:
```diff
@@ -140,6 +140,12 @@ class BaseExpression:
 
+    # If the expression can be used in a WHERE-clause.
+    filterable = True
+
```
I wouldn't include blank lines between the attributes since they already have comments separating them.

timgraham commented on django/db/models/expressions.py:
```diff
@@ -140,6 +140,12 @@ class BaseExpression:
     is_summary = False
     _output_field_resolved_to_none = False
 
+    # If the expression can be used in a WHERE-clause.
```
"Can the expression be used in a WHERE-clause?"

atombrella replied:
I think it was already commented a few times, so make yours final? :)

timgraham commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,125 @@
+
+
+class CumeDist(Func):
+    window_compatible = True
```
alphabetize that below function/name?

timgraham commented on django/db/models/sql/compiler.py:
```diff
@@ -1110,6 +1110,8 @@ def prepare_value(self, field, value):
             if value.contains_aggregate:
                 raise FieldError("Aggregate functions are not allowed in this query")
+            if value.contains_over_clause:
+                raise FieldError('Window expressions are not allowed in this query.')
```
I only see 1 test for "Window expressions are not allowed in this query." but there are two places in this file with that message. Is a test missing?

atombrella replied:
Probably. However, there's already `F() expressions can only be used to update, not to insert.` which catches the typical use of this expression. How about this?

```
def test_fail_insert(self):
    """Window expressions can't be used in an INSERT statement."""
    msg = 'Window expressions are not allowed in this query'
    with self.assertRaisesMessage(FieldError, msg):
        Employee.objects.create(
            name='Jameson', department='Management', hire_date=datetime.date(2007, 7, 1),
            salary=Window(expression=Sum(Value(10000))),
        )
```

timgraham commented on django/db/models/window.py:
```diff
@@ -0,0 +1,186 @@
+from django.db import connection
```
Maybe `expressions.py` should be converted into a package and this moved into there?

atombrella replied:
Move the new expressions to this file, and reorganize later?

timgraham commented on django/db/models/window.py:
```diff
@@ -0,0 +1,186 @@
+    template = '%(expression)s OVER (%(window)s)'
+    filterable = False
+
+    # although the main expression may either be an aggregate or an
```
use capitals/periods

timgraham commented on django/db/models/window.py:
```diff
@@ -0,0 +1,186 @@
+
+    def as_sql(self, compiler, connection):
+        """
+        This should be in operations for each of the database backend that
```
This docstring reads as a TODO. Can you clarify the meaning (i.e. what is "This")?

atombrella replied:
I think the comment could be removed.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -819,3 +819,146 @@ Usage example::
+
+.. versionadded:: 2.0
+
+In addition to aggregate functions, e.g., ``Sum`` and ``Count``, there is a
```
there are a number

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -819,3 +819,146 @@ Usage example::
+In addition to aggregate functions, e.g., ``Sum`` and ``Count``, there is a
+number of functions to use in a window expression
+:class:`~django.db.models.expressions.Window` for computing the rank of
+elements, computing the :class:`Ntile` of some rows. Some of these functions do
```
missing a conjunction where the comma is?

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -819,3 +819,146 @@ Usage example::
+
+.. class:: RowNumber(*expressions)
+
+    Will compute the row number according to the ordering of either the frame
```
Will compute -> Computes

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,801 @@
+            ('Wilkinson', 60000, 'IT', None),
+            ('Johnson', 80000, 'Management', None),
+            ('Miller', 100000, 'Management', None),
+        ], lambda row: (row.name, row.salary, row.department, row.nth_value))
```
It seems that ordering is an issue, therefore:
```python
qs = Employee.objects.annotate(nth_value=Window(
    expression=NthValue('salary', nth=20),
    order_by=F('salary').asc()
))
self.assertListEqual(list(qs.values_list('nth_value', flat=True).distinct()), [None])
```
will work properly.

atombrella replied:
Thanks :)

atombrella commented:
I went through the feedback, but I still find the documentation tricky to get finalized (I fixed a few sloppy mistakes in the previous updates; it's squashed). Regarding examples, should I add sample data as well, or the description of what is enough? I think for some of those queries to make sense and fully
illustrate what happens would require more than 4-5 rows (that's the number of in the aggregation section), making the document much longer. I use movies and books in three examples. The conditional aggregation document fully illustrates with a model, and some data, for comparison. All the documentation is now in already existing files.

I believe all scenarios for the functions have tests, however there's a test, `test_two_functions` to check how two window expressions work. Perhaps it's redundant? There are some assertions to check the generated SQL, mainly for the frames-tests.

atombrella commented:
buildbot, test on oracle.

atombrella commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,806 @@
+    def test_invalid_start_value_range(self):
+        msg = "start argument must be a negative integer, zero, or None, but got '3'."
+        with self.assertRaisesMessage(ValueError, msg):
+            qs = Employee.objects.annotate(test=Window(
```
Yes, it also seems cleaner/nicer. `qs[0]` was just to trigger evaluation of the queryset.

atombrella commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,125 @@
+    name = 'Ntile'
+    window_compatible = True
+
+    def __init__(self, num_buckets=1, **extra):
```
I think it should always be passed. I'll update along with the failing tests from Oracle.

atombrella commented:
buildbot, test on oracle.

timgraham commented on django/db/models/functions/window.py:
```diff
@@ -0,0 +1,125 @@
+    window_compatible = True
+
+    def __init__(self, *expressions, **extra):
+        super().__init__(*expressions, output_field=FloatField(), **extra)
```
You can update this after 08654a99bbdd09049d682ae57cc94241534b29f0, I believe.

atombrella replied:
Yes, I just noticed after I did a `pull` from upstream.

timgraham commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+
+    .. code-block:: python
+
+        >>> from django.db.models import Avg, ExpressionList, F, Window
```
`ExpressionList` is imported but not used. Does it need to be a public API? It seems only used internally to `Window`. If it has other uses, could it be added in a separate PR to decrease the size of this one?

atombrella replied:
I'd be happy to make it private API (would it be enough to just remove the docs, and maybe make it an internal class?), and if someone wants it, we can move it out from there. It was a really convenient way to make it more flexible to pass many arguments to `order_by` or `partition`.

timgraham replied:
Removing the docs is enough.

timgraham commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+``RANGE``, the frame starts or ends at the first or last peer according to the
+ordering-clause. Thus, ``RANGE CURRENT ROW`` evaluates the expression for rows
+which have the same value specified by the ordering. Because the template is
+specific, this may be specified with::
```
I'm not sure what "the template is specific" means -- is it a typo?

atombrella replied:
For instance SQL allows `RANGE CURRENT ROW`, but the template includes is set to include both the start and end point which might be a bit misleading.

timgraham commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+      >>>     avg_rating=Window(
+      >>>         expression=Avg('rating'),
+      >>>         partition_by=[F('studio'), F('genre')],
+      >>>         order_by=F('released__year').asc(),
```
Seems like this would only sort movies by year... would month/day ordering be unspecified then? Doesn't seem like that's what we want.

atombrella replied:
It should be `ExtractYear` (or `TruncYear`).

atombrella commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+</topics/db/aggregation>` method. The ORM does not make use of named windows,
+instead they are part of the selected columns. If you want to apply multiple
+expressions over the same window, e.g., the same partition and frame, it's
+convenient to use a dictionary and pass it as an argument with ``**params``.
```
Yes. I'll add something.

atombrella commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+
+.. class:: ValueRange(start=None, end=None)
+
+    .. attribute:: frametype
```
Yes, changed.

atombrella commented on docs/ref/models/database-functions.txt:
```diff
@@ -819,3 +819,147 @@ Usage example::
+
+.. note::
+
+    PostgreSQL does not support ``RESPECT NULLS`` or ``IGNORE NULLS`` option
```
I'm not sure if this actual belongs in the documentation of Django, it's mostly rewritten from the documentation in PostgreSQL, and I think users will already know this pitfalls. It's highlighted in PostgreSQL's documentation.

atombrella commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+
+        The SQL standard disallows referencing the window function in the
+        ``WHERE``-clause and Django raises an exception if it's attempted.
+        This scenario may be solved with subqueries. Defaults to ``False``.
```
The subquery sentence should be removed. https://code.djangoproject.com/ticket/28333 is created for the same reason.

atombrella commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+    For example, to add an extra column that has the accumulated average for
+    the movies in the same genre and released in the year. Thus, it's easy to
+    see if a movie is rated better or worse than the average. This is done per
+    movie, and not as a total.
```
I see you removed it in the edits commit. Thanks! :)

atombrella commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +666,185 @@ should avoid them if possible.
+Frames are used for narrowing which rows that should be used for computing the
+result. To ease the understanding of frames, they shift from some start point
+to some specified end point. Frames can be used with and without partitions,
+but it is recommended to specify an ordering of the window. In a frame, a peer
```
Mainly just to have to have deterministic results. For instance, for the test data `Rank` over a blank window produces 1, 1, 1, 1 but with an ordering of `salary`, it changes. It's also not given that `order_by` is optional in a database backend, e.g., SQL server and Oracle require it.

atombrella commented:
@timgraham I went over the feedback, and the coverage should be increased. I don't know if `BackendsOperationsTests` actually makes sense. I decided to delete the `admonition` about ordering and frames; I think the content is described already.

atombrella commented:
buildbot, test on oracle.

felixxm commented on django/db/models/expressions.py:
```diff
@@ -1118,3 +1151,168 @@ def asc(self):
+        }
+
+    def window_frame_start_end(self, connection, start, end):
+        raise NotImplemented('Subclasses must implement this.')
```
i think you should use `NotImplementedError` instead on `NotImplemented`, moreover this proves that it's untested 😞.

atombrella replied:
Yes, d'oh. Fixed.

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,795 @@
+        qs = Employee.objects.annotate(min_salary=Window(
+            expression=Min('salary'),
+            partition_by=F('department'),
+            order_by=[F('salary').asc(), F('name').asc()])
```
The last parenthesis should be moved to the next line due to hanging indentation.

atombrella replied:
Thanks!

felixxm commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,795 @@
+        qs = Employee.objects.annotate(max_salary_year=Window(
+            expression=Max('salary'),
+            order_by=ExtractYear('hire_date').asc(),
+            partition_by=ExtractYear('hire_date'))
```
The last parenthesis should be moved to the next line due to hanging indentation.

atombrella commented on tests/expressions_window/tests.py:
```diff
@@ -0,0 +1,800 @@
+
+    def test_window_frame_repr(self):
+        row_frame = RowRange(start=-1)
+        self.assertEqual(repr(row_frame), '<RowRange: ROWS BETWEEN {} AND {}>'.format(
```
I put everything (`repr` and validation of the expressions) that doesn't directly involve a call to the database in `NonQueryWindowTest`. Maybe you have a better name for the testclass?

atombrella commented on docs/ref/models/expressions.txt:
```diff
@@ -677,6 +862,30 @@ calling the appropriate methods on the wrapped expression.
+
+        Tells Django that this expression contains a
+        :class:`~django.db.models.expressions.Window` expression. Queries with a
+        window function expression are mainly disallowed in queries that
```
You can't use them in `INSERT` or `UPDATE` (at least not in Django, I think a work-around in SQL is common table expressions although I've never tried), and PostgreSQL has a caveat with `FOR UPDATE/SHARE`. I used "mainly" because there may be other queries where an error is raised by the database. I think I'm more keen to delete "usually", and try to replace with something like "For instance, window function expressions can't be used `INSERT` or `UPDATE` queries".

atombrella commented on docs/ref/models/expressions.txt:
```diff
@@ -649,6 +656,184 @@ should avoid them if possible.
+equivalent to including multiple expressions in a :doc:`QuerySet.annotate()
+</topics/db/aggregation>` call. The ORM doesn't make use of named windows,
+instead they are part of the selected columns. If you want to apply multiple
+expressions over the same window, e.g., the same partition and frame, it's
```
There's the also the ordering, which would handle what's a peer and what's not a peer; I don't know if it's implied. I think "i.e." and "e.g." are both fine to use.

atombrella commented:
@timgraham I think the rest of feedback is addressed.

adamchainz commented:
🎉  Awesome 🎉

atombrella commented:
Thanks to people involved in getting this mergeable :tada:
